statsmodels/base/model.py

```diff
@@ -510,6 +510,164 @@ def hess(params, *args):
+        keep_index : array_like (int or bool) or slice
+            variables that should be dropped.
+        start_params : None or array_like
+            starting values for the optimization. `start_params` needs to be
+            given in the original parameter space and are internally
+            transformed.
+        k_params : int or None
+            If None, then we try to infer from start_params or model.
+        **fit_kwds : keyword arguments
+            fit_kwds are used in the optimization of the transformed model.
+
+        Returns
+        -------
+        results : Results instance
+
+        """
+        # we need to append index of extra params to keep_index as in
+        # NegativeBinomial
+        if hasattr(self, 'k_extra') and self.k_extra > 0:
+            # we cannot change the original, TODO: should we add keep_index_params?
+            keep_index = np.array(keep_index, copy=True)
+            k = self.exog.shape[1]
+            extra_index = np.arange(k, k + self.k_extra)
+            keep_index_p = np.concatenate((keep_index, extra_index))
+        else:
+            keep_index_p = keep_index
+
+        # not all models support start_params, drop if None, hide them in fit_kwds
+        if start_params is not None:
+            fit_kwds['start_params'] = start_params[keep_index_p]
+            k_params = len(start_params)
+            # ignore k_params in this case, or verify consisteny?
+
+        # build auxiliary model and fit
+        init_kwds = self._get_init_kwds()
+        mod_constr = self.__class__(self.endog, self.exog[:, keep_index],
+                                    **init_kwds)
+        res_constr = mod_constr.fit(**fit_kwds)
+        # switch name, only need keep_index for params below
+        keep_index = keep_index_p
+
+        if k_params is None:
+            k_params = self.exog.shape[1]
+            k_params += getattr(self, 'k_extra', 0)
+
+        params_full = np.zeros(k_params)
+        params_full[keep_index] = res_constr.params
+
+        # create dummy results Instance, TODO: wire up properly
+        # TODO: this could be moved into separate private method if needed
+        # discrete L1 fit_regularized doens't reestimate AFAICS
+        # RLM doesn't have method, disp nor warn_convergence keywords
+        # OLS, WLS swallows extra kwds with **kwargs, but doesn't have method='nm'
+        try:
+            # Note: addding full_output=False causes exceptions
+            res = self.fit(maxiter=0, disp=0, method='nm', skip_hessian=True,
+                           warn_convergence=False, start_params=params_full)
+            # we get a wrapper back
+        except (TypeError, ValueError):
+            res = self.fit()
+
+        # Warning: make sure we are not just changing the wrapper instead of
+        # results #2400
+        if hasattr(res_constr.model, 'scale'):
+            # GLM problem, see #2399
+            res.model.scale = res._results.scale = res_constr.model.scale
```

Comment:
Is `res.model` not `self`?

(tangential to this PR) I know the GLM model de-facto has a `scale` attribute, but does it actually need it?  I'm under the impression that it only really _belongs_ on the GLMResults instance.

Reply:
> Is res.model not self?

Yes, but I hadn't realized it, I leave it but added a code comment

GLM is in 2399 cited on the line above